File: SEPTEMBER-2025/02-09-2025/p3025.cpp
#define vi vector<int>
#define vvi vector<vi>
class Solution {
public:
    int numberOfPairs(vector<vector<int>>& points) {
        int n = points.size();
        // Sort points: primarily by x (ascending), and for same x, by y (descending)
        // This ensures we process points from left to right, and for same x from top to bottom
        sort(points.begin(), points.end(), [&](auto& x, auto& y) -> bool {
            return x[0] == y[0] ? x[1] > y[1] : x[0] < y[0];
        });

        int ans = 0;
        // Iterate over each point as candidate A (upper left)
        for (int i = 0; i < n; ++i) {
            int a = points[i][1];  // y-coordinate of point A
            int last = -1;  // Tracks the highest y-coordinate encountered so far in this iteration (initialized to -1 as lowest)
            
            // For each point j to the right of i, check if it can be candidate B (lower right)
            for (int j = i + 1; j < n; ++j) {
                int b = points[j][1];  // y-coordinate of point B
                // Check if A is above B (a >= b) AND 
                // if current point B's y is greater than 'last' (meaning no point between them has a y between b and a)
                if (a >= b and last < b) {
                    last = b;  // Update last to current y
                    ++ans;     // Valid pair found
                }
            }
        }

        return ans;
    }
};

File: SEPTEMBER-2025/02-09-2025/p3025.java
class Solution {
    public int numberOfPairs(int[][] points) {
         int n = points.length;

        // Step 1: Sort by x increasing, and if tie, y decreasing
        Arrays.sort(points, (a, b) -> {
            if (a[0] != b[0]) return a[0] - b[0];
            return b[1] - a[1];
        });

        int ans = 0;

        // Step 2: Nested scan with maxY check
        for (int i = 0; i < n; i++) {
            int maxY = Integer.MIN_VALUE;
            for (int j = i + 1; j < n; j++) {
                if (points[j][1] <= points[i][1] && points[j][1] > maxY) {
                    ans++;
                    maxY = points[j][1];
                }
            }
        }

        return ans;
        
    }
}